qiita_db/handlers/oauth2.py

```diff
@@ -0,0 +1,280 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) 2014--, The Qiita Development Team.
+#
+# Distributed under the terms of the BSD 3-clause License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# -----------------------------------------------------------------------------
+
+from base64 import urlsafe_b64decode
+from string import ascii_letters, digits
+import datetime
+from random import SystemRandom
+import functools
+
+from tornado.web import RequestHandler
+
+from moi import r_client
+from qiita_core.exceptions import (IncorrectPasswordError, IncorrectEmailError,
+                                   UnverifiedEmailError)
+import qiita_db as qdb
+
+
+def login_client(client_id, client_secret=None):
+    """Login client_id, optionally with client_secret
+
+    Parameters
+    ----------
+    client_id : str
+        The client ID to validate
+    client_secret : str, optional
+        The client secret code. If not given, will only validate if client_id
+        exists
+
+    Returns
+    -------
+    bool
+        Whether client_id exists and, if client_secret given, the secret
+        matches the id
+    """
+    with qdb.sql_connection.TRN:
+        sql = """SELECT EXISTS(
+                    SELECT *
+                    FROM qiita.oauth_identifiers
+                    WHERE client_id = %s{0})"""
+        sql_info = [client_id]
+        if client_secret is not None:
+            sql = sql.format("AND client_secret = %s")
+            sql_info.append(client_secret)
+        else:
+            sql = sql.format("AND client_secret IS NULL")
+        qdb.sql_connection.TRN.add(sql, sql_info)
+        return qdb.sql_connection.TRN.execute_fetchlast()
+
+
+def authenticate_oauth(f):
+    """Decorate methods with this to require valid Oauth2 Authorization header
+
+    If a valid header is given, teh handof is odne and the page is rendered. If
+    an invalid header is given, the json error message is automatically sent.
+    """
+    @functools.wraps(f)
+    def wrapper(self, *args, **kwargs):
+        def oauth_error(self, error_msg, error):
```

Comment:
It looks like the code is written so if you need oauth2 you need to extend from OauthBaseHandler, e.g. you're calling check_rate_limit below.